### PR TITLE
Task/aca portfolio 60

### PR DESCRIPTION
**Pull request by paola-pc**
Development of AcaPortfolio 2.0
Depends on https://github.com/leemonade/bubbles/pull/132

### CODE REVIEW

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateKnowledgeArea.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import { allKnowledgeAreaKeys, getKnowledgeAreasKey } from '../keys/knowledgeAreas';
+
+export function useCreateKnowledgeArea() {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: async (props) => createKnowledgeAreaRequest(props),
+    onSuccess: (data) => {
+      // Invaidate knowledge-areas query for that center
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateKnowledgeArea.js:15]:
```suggestion
      // Invalidate knowledge-areas query for that center
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateProgram.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    },
+  });
+}
+
+export function useCreateProgram() {
+  const queryClient = useQueryClient();
+  return useMutation({
+    mutationFn: async (props) => createProgramRequest(props),
+    onSuccess: (data) => {
+      // Invaidate centerPrograms query
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateProgram.js:36]:
```suggestion
      // Invalidate centerPrograms query
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateProgram.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    },
+  });
+}
+
+export function useDuplicateProgram() {
+  const queryClient = useQueryClient();
+  return useMutation({
+    mutationFn: async (props) => duplicateProgramRequest(props),
+    onSuccess: (data) => {
+      // Invaidate centerPrograms query
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateProgram.js:51]:
```suggestion
      // Invalidate centerPrograms query
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateSubjectType.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import { getSubjectTypesKey } from '../keys/subjectTypes';
+
+export function useUpdateSubjectType() {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: async (props) => updateSubjectTypeRequest(props),
+    onSuccess: (data) => {
+      // Invaidate subject-types query for that center
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateSubjectType.js:15]:
```suggestion
      // Invalidate subject-types query for that center
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateSubjectType.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  });
+}
+
+export function useCreateSubjectType() {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: async (props) => createSubjectTypeRequest(props),
+    onSuccess: (data) => {
+      // Invaidate subject-types query for that center
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/mutations/useMutateSubjectType.js:28]:
```suggestion
      // Invalidate subject-types query for that center
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/pages/private/AcademicTreePage.js [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+    }
+    return [];
+  }, [centerProgramsQuery]);
+
+  const { data: academicTreeQuery, isLoading: isAcademicTreeLoading } = useProgramAcademicTree({
+    programId: selectedProgram,
+    options: { enabled: selectedProgram?.length > 0 },
+  });
+
+  const generateGUID = () => `_${Math.random().toString(36).substr(2, 9)}`;
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/pages/private/AcademicTreePage.js:70]: Watch out, this is not unique, I recommend to use `uuid`

File: plugins/leemons-plugin-assignables/frontend/src/components/Assignment/AssignStudents/hooks/useGroupedClasses.js [hunk 7]
```diff
@@ -89,6 +89,7 @@ export default function useGroupedClasses(subjects, disableGrouping = false) {
     }, []);
 
     const students = _.uniq(groups.flatMap((c) => c.students));
+    // console.log('groups', groups)
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-assignables/frontend/src/components/Assignment/AssignStudents/hooks/useGroupedClasses.js:92]:
```suggestion
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/centerPrograms.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+export const allCenterPrograms = [
+  {
+    plugin: 'plugin.academic-portfolio',
+    scope: 'center-programs',
+  },
+];
+
+export const getCenterProgramsKey = (center, filters) => [
+  {
+    ...allCenterPrograms,
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/centerPrograms.js:10]:
```suggestion
    ...allCenterPrograms[0],
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/programCourse.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+export const allCourseDetailKeys = [
+  {
+    plugin: 'plugin.academic-portfolio',
+    scope: 'course-detail',
+  },
+];
+
+export const getCourseDetailKey = (id) => [
+  {
+    ...allCourseDetailKeys,
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/programCourse.js:10]:
```suggestion
    ...allCourseDetailKeys[0],
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/programGroup.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+export const allGroupDetailKeys = [
+  {
+    plugin: 'plugin.academic-portfolio',
+    scope: 'group-detail',
+  },
+];
+
+export const getGroupDetailKey = (id) => [
+  {
+    ...allGroupDetailKeys,
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/programGroup.js:10]:
```suggestion
    ...allGroupDetailKeys[0],
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/programSubjects.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+export const allProgramSubjectsKeys = [
+  {
+    plugin: 'plugin.academic-portfolio',
+    scope: 'program-subjects',
+  },
+];
+
+export const getProgramSubjectsKey = (program, filters) => [
+  {
+    ...allProgramSubjectsKeys,
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/programSubjects.js:10]:
```suggestion
    ...allProgramSubjectsKeys[0],
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/programTree.js [hunk 12]
```diff
@@ -0,0 +1,13 @@
+export const allProgramTreeKeys = [
+  {
+    plugin: 'plugin.academic-portfolio',
+    scope: 'program-tree',
+  },
+];
+
+export const getProgramTreeKey = (program) => [
+  {
+    ...allProgramTreeKeys,
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/programTree.js:10]:
```suggestion
    ...allProgramTreeKeys[0],
```

File: plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/subjectTypes.js [hunk 13]
```diff
@@ -0,0 +1,13 @@
+export const allSubjectTypesKeys = [
+  {
+    plugin: 'plugin.academic-portfolio',
+    scope: 'subject-types',
+  },
+];
+
+export const getSubjectTypesKey = (center) => [
+  {
+    ...allSubjectTypesKeys,
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-academic-portfolio/frontend/src/hooks/keys/subjectTypes.js:10]:
```suggestion
    ...allSubjectTypesKeys[0],
```

File: plugins/leemons-plugin-grades/frontend/src/hooks/keys/centerEvaluationSystems.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
+export const allCenterEvaluationSystemsKeys = [
+  {
+    plugin: 'plugin.grades',
+    scope: 'center-evaluation-systems',
+  },
+];
+
+export const getCenterEvaluationSystemsKey = (center) => [
+  {
+    ...allCenterEvaluationSystemsKeys,
```
Comment (MIGUELez11) [inline plugins/leemons-plugin-grades/frontend/src/hooks/keys/centerEvaluationSystems.js:10]:
```suggestion
    ...allCenterEvaluationSystemsKeys[0],
```

**Review by MIGUELez11 — CHANGES_REQUESTED** [review]
Well, what a PR, very nice code, I love pretty much of it, what a huge improvement of the Academic Portfolio code!!!! 👏🏻

Please review the suggestions and comments and tomorrow we review it together, also take a look on the useQuery hooks, as most of them are not passing all the variables that change the request output to the key, which will make some requests to not work properly.

File: plugins/leemons-plugin-academic-portfolio/frontend/src/pages/private/AcademicTreePage.js (hunk 6 above)
Comment (paola-pc) [reply plugins/leemons-plugin-academic-portfolio/frontend/src/pages/private/AcademicTreePage.js]: Hecho!

**Review by MIGUELez11 — APPROVED** [review]

**Pull request merged**